src/utils/form.ts

```diff
@@ -12,3 +14,148 @@ export const hiddenInputStyle = `
   -webkit-appearance: none !important;
   z-index: -1 !important;
 `;
+
+interface FormDataEvent extends Event {
+  formData: FormData;
+}
+
+/**
+ * Defines interface for components that are form-compatible.
+ *
+ * Along with the interface, use the matching form utils to help set up the component behavior.
+ */
+export interface FormAssociated<T = any> {
+  /**
+   * The host element.
+   */
+  readonly el: HTMLElement;
+
+  /**
+   * The form this component is associated with.
+   */
+  formEl: HTMLElement;
+
+  /**
+   * The name used to submit the value to the associated form.
+   *
+   * Note that this prop should use the @Prop decorator.
+   *
+   * @default ""
+   */
+  name: string;
+
+  /**
+   * This form component's value.
+   *
+   * Note that this prop should use the @Prop decorator.
+   */
+  value: T;
+
+  /**
+   * The initial value for this form component.
+   *
+   * When the form is reset, the value will be set to this property.
+   */
+  initialValue: T;
+
+  /**
+   * Hook for components to provide custom form reset behavior.
+   */
+  onFormReset?(): void;
+}
+
+const onFormDataMap = new WeakMap<HTMLElement, typeof onFormData>();
+const onFormResetMap = new WeakMap<HTMLElement, FormAssociated["onFormReset"]>();
+const hiddenInputMap = new WeakMap<HTMLElement, HTMLInputElement>();
+
+// Safari does not support `formdata` event
+const supportsFormDataEvent = "FormDataEvent" in window;
+
+/**
+ * Helper to set up form interactions on connectedCallback.
+ *
+ * @param el - the host element
+ */
+export function connectForm<T>(formAssociated: FormAssociated<T>): void {
```

Comment:
I like how you pass the instance here. I'll update my `connectLabel` and `disconnectLabel` to do the same.